wireapp/kalium
# fix: wait for logout to be completed when handling user events [WPB-9689] [WPB-10310]

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-9689" title="WPB-9689" target="_blank"><img alt="Bug" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10803?size=medium" />WPB-9689</a>  [Android] App crashing and stuck on splash screen when deleting device for second account from another client
  </td></table>
  <br />
 






----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

When the app is in the background and the current client is removed from other client, then when launching the app there's a crash.

### Causes (Optional)

`ProteusPreKeyRefiller` always launches when sync completes (when incremental sync state is changed to `Live`) and it 
requires core crypto to generate new pre keys, but when the app received `User.ClientRemove` event and started handling it as an "offline event" (before sync is `Live`) then core crypto may be removed before or during executing the refiller. 
Generally I noticed three possible outcomes:
- fatal signal error (generating new prekeys already started and core crypto got removed during that)
- ProteusException: Job was cancelled (probably userscope got removed and cancelled all ongoing coroutines)
- ProteusException: CoreCrypto object has already been destroyed (core crypto got removed before some prekeys started being generated)
Just the first one ends up crashing the app, `ProteusException`s are handled quietly and don't crash the app, so depending on what happens first, it can crash or handle it without a crash.

### Solutions

We shouldn't start refilling pre keys if the core crypto / client is already removed. 
Even more - we shouldn't execute any action after sync or handle any other event after receiving `User.ClientRemove` or `User.UserDelete` event. 
The solution is easy - add `waitUntilCompletes = true` when when executing `logout` during handling `User.ClientRemove` or `User.UserDelete` event. This way, when the app starts the logout, it should actually wait for the logout to be completed before any other event is handled, so it will close the user session and related coroutine scope and prevent the app from handling subsequent events and doing any actions after the sync is completed.

### Testing

#### Test Coverage (Optional)


#### How to Test

- login on client A
- login as the same user on client B
- put client A into background
- remove client A on client B
- open the app on client A by clicking on the app icon
- user should be logged out on client A without any crash

----
#### PR Post Submission Checklist for internal contributors (Optional)

----
#### PR Post Merge Checklist for internal contributors

----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.
